Clarify cluster diagram as one possible reference architecture (#47164)
kubernetes/website: content/en/docs/concepts/architecture/_index.md

Reviewer: Even better: migrate in **all** the text content from https://k8s.io/docs/concepts/overview/components/, and at the same time make https://k8s.io/docs/concepts/overview/components/ be a much simpler overview. This page becomes the detailed explanation, and https://k8s.io/docs/concepts/overview/components/ becomes the overview.

You can also explain in _this_ page that there is more than one way to lay out a control plane and that running the control plane inside containers is just one viable option.

:information_source: The feedback in this comment is **optional**. But I do recommend doing that switcheroo; it's what the docs really need.

Author: I think this is a great suggestion, thanks! I'll work on refactoring this along those lines.

Author: I have drafted a change along these lines and added it in the latest commit, let me know if there is anything I can change! I wasn't too sure about my added section on `Architecture Variations`

  The architectural concepts behind Kubernetes.
---

> A Kubernetes cluster consists of a set of worker machines, called nodes, that run containerized applications. Every cluster has at least one worker node.

Reviewer: You can now have a zero-node cluster and scale out on demand, so:
```suggestion
A Kubernetes cluster consists of a control plane plus a set of worker machines, called nodes, that run containerized applications. Every cluster needs at least one worker node in order to run Pods.
```

The worker node(s) host the Pods that are the components of the application workload. The control plane manages the worker nodes and the Pods in the cluster. In production environments, the control plane usually runs across multiple computers and a cluster usually runs multiple nodes, providing fault-tolerance and high availability.

This document outlines the various components you need to have for a complete and working Kubernetes cluster.

{{< figure src="/images/docs/kubernetes-cluster-architecture.svg" alt="The control plane (kube-apiserver, etcd, kube-controller-manager, kube-scheduler) and several nodes. Each node is running a kubelet and kube-proxy."
title="Kubernetes cluster components"
caption="**Note:** This diagram presents an example reference architecture for a Kubernetes cluster. The actual distribution of components can vary based on specific cluster setups and requirements." class="diagram-large" >}}

> ## Control Plane Components

Reviewer: (nit / optional fix)
```suggestion
## Control plane components
```

- ServiceAccount controller: Create default ServiceAccounts for new namespaces.

The above is not an exhaustive list.

### cloud-controller-manager

{{< glossary_definition term_id="cloud-controller-manager" length="short" >}}

The cloud-controller-manager only runs controllers that are specific to your cloud provider. If you are running Kubernetes on your own premises, or in a learning environment inside your own PC, the cluster does not have a cloud controller manager.

As with the kube-controller-manager, the cloud-controller-manager combines several logically independent control loops into a single binary that you run as a single process. You can scale horizontally (run more than one copy) to improve performance or to help tolerate failures.

The following controllers can have cloud provider dependencies:

- Node controller: For checking the cloud provider to determine if a node has been deleted in the cloud after it stops responding
- Route controller: For setting up routes in the underlying cloud infrastructure
- Service controller: For creating, updating and deleting cloud provider load balancers

> ## Node Components

Reviewer: (nit / optional fix)
```suggestion
## Node components
```

Node components run on every node, maintaining running pods and providing the Kubernetes runtime environment.

### kubelet

{{< glossary_definition term_id="kubelet" length="all" >}}

> ### kube-proxy

Reviewer: Optional improvement
```suggestion
### kube-proxy (optional) {#kube-proxy)
```

{{< glossary_definition term_id="kube-proxy" length="all" >}}

> 

Reviewer: Optional improvement
```suggestion
If you use a [network plugin](#network-plugins) that implements packet forwarding for Services
by itself, and providing equivalent behavior to kube-proxy), then you do not need to run
kube-proxy on the nodes in your cluster.

```

### Container runtime

{{< glossary_definition term_id="container-runtime" length="all" >}}

> ## Addons
> 
> Addons use Kubernetes resources ({{< glossary_tooltip term_id="daemonset" >}}, {{< glossary_tooltip term_id="deployment" >}}, etc) to implement cluster features. Because these are providing cluster-level features, namespaced resources for addons belong within the `kube-system` namespace.
> 
> Selected addons are described below; for an extended list of available addons, please see [Addons](/docs/concepts/cluster-administration/addons/).

Reviewer 2: ```suggestion
## Add-ons

Add-ons use Kubernetes resources ({{< glossary_tooltip term_id="daemonset" >}}, {{< glossary_tooltip term_id="deployment" >}}, etc) to implement cluster features. Because these are providing cluster-level features, namespaced resources for add-ons belong within the `kube-system` namespace.

Selected add-ons are described below; for an extended list of available add-ons, please see [Addons](/docs/concepts/cluster-administration/addons/).
```
It's easy to confuse [add-on](https://en.wikipedia.org/wiki/Add-on) with [addon](https://en.wikipedia.org/wiki/Addon).

Reviewer 2: This suggestion is debatable since the file name here is addons, `content/en/docs/concepts/cluster-administration/addons.md`.
I'll be happy to open a PR to fix the file name. What do you think, @sftim?

Reviewer: I think this text may have been moved rather than written new. We avoid providing much corrective feedback to the parts of a change where a contributor is only moving text around.

Author: Thank you for the feedback @network-charles!

Given that this text was moved from an existing file and to maintain consistency with current documentation, I suggest keeping the term 'addons' without the hyphen. If a broader change to 'add-ons' is desired across the documentation, that might be better addressed in a separate, dedicated PR.

Reviewer 2: Oh, ok. I thought the text was new. Yes, I'll open a new PR addressing the 'addons' issue when this PR has been merged.

Author: Sounds like a great idea, @network-charles. I appreciate your attention to detail.

### DNS

> While the other addons are not strictly required, all Kubernetes clusters should have [cluster DNS](/docs/concepts/services-networking/dns-pod-service/), as many examples rely on it.

Reviewer 2: ```suggestion
While the other add-ons are not strictly required, all Kubernetes clusters should have a [cluster DNS](/docs/concepts/services-networking/dns-pod-service/), as many examples rely on it.
```

Author: @network-charles I'll skip this change to maintain consistency with current docs as discussed above. Thanks for flagging it.

Cluster DNS is a DNS server, in addition to the other DNS server(s) in your environment, which serves DNS records for Kubernetes services.

Containers started by Kubernetes automatically include this DNS server in their DNS searches.

### Web UI (Dashboard)

[Dashboard](/docs/tasks/access-application-cluster/web-ui-dashboard/) is a general purpose, web-based UI for Kubernetes clusters. It allows users to manage and troubleshoot applications running in the cluster, as well as the cluster itself.

> ### Container Resource Monitoring

Reviewer: (nit / optional fix)
```suggestion
### Container resource monitoring
```

[Container Resource Monitoring](/docs/tasks/debug/debug-cluster/resource-usage-monitoring/) records generic time-series metrics about containers in a central database, and provides a UI for browsing that data.

### Cluster-level Logging

> A [cluster-level logging](/docs/concepts/cluster-administration/logging/) mechanism is responsible for saving container logs to a central log store with search/browsing interface.

Reviewer 2: ```suggestion
A [cluster-level logging](/docs/concepts/cluster-administration/logging/) mechanism is responsible for saving container logs to a central log store with a search/browsing interface.
```

> ### Network Plugins

Reviewer: (nit / optional fix)
```suggestion
### Network plugins
```

[Network plugins](/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins) are software components that implement the container network interface (CNI) specification. They are responsible for allocating IP addresses to pods and enabling them to communicate with each other within the cluster.

> ## Architecture Variations

Reviewer: (nit / optional fix)
```suggestion
Architecture variations
```

Author: Was the deletion of `##` intentional here? I feel this text should remain as a heading to maintain consistency with the other sections.

Author: I've kept the heading but fixed the casing, feel free to fix it if that was what was intended, thanks!

While the core components of Kubernetes remain consistent, the way they are deployed and managed can vary. Understanding these variations is crucial for designing and maintaining Kubernetes clusters that meet specific operational needs.

> ### Control Plane Deployment Options

Reviewer: (nit / optional fix)
```suggestion
### Control plane deployment options
```

The control plane components can be deployed in several ways:

> <dl>
>   <dt>Traditional deployment</dt>
>   <dd>Control plane components run directly on dedicated machines or VMs, often managed as systemd services.</dd>
> 
>   <dt>Static Pods</dt>
>   <dd>Control plane components are deployed as static Pods, managed by the kubelet on specific nodes. This is a common approach used by tools like kubeadm.</dd>
> 
>   <dt>Self-hosted</dt>
>   <dd>The control plane runs as Pods within the Kubernetes cluster itself, managed by Deployments or other Kubernetes primitives.</dd>
> 
>   <dt>Managed Kubernetes services</dt>
>   <dd>Cloud providers often abstract away the control plane, managing its components as part of their service offering.</dd>
> </dl>

Reviewer: (nit / optional fix)
```suggestion
Traditional deployment
: Control plane components run directly on dedicated machines or VMs, often managed as systemd services.

Static Pods
: Control plane components are deployed as static Pods, managed by the kubelet on specific nodes. This is a common approach used by tools like kubeadm.

Self-hosted
: The control plane runs as Pods within the Kubernetes cluster itself, managed by Deployments and StatefulSets or other Kubernetes primitives.

Managed Kubernetes services
: Cloud providers often abstract away the control plane, managing its components as part of their service offering.
```

Author: I wasn't aware you can write definition lists like this, thanks!

> ### Workload Placement Considerations

Reviewer: (nit / optional fix)
```suggestion
### Workload placement considerations
```

The placement of workloads, including the control plane components, can vary based on cluster size, performance requirements, and operational policies:

- In smaller or development clusters, control plane components and user workloads might run on the same nodes.
- Larger production clusters often dedicate specific nodes to control plane components, separating them from user workloads.
- Some organizations run critical add-ons or monitoring tools on control plane nodes.

> ### Cluster Management Tools

Reviewer: (nit / optional fix)
```suggestion
### Cluster management tools
```

Tools like kubeadm, kops, and Kubespray offer different approaches to deploying and managing clusters, each with its own method of component layout and management.

The flexibility of Kubernetes architecture allows organizations to tailor their clusters to specific needs, balancing factors such as operational complexity, performance, and management overhead.

> ### Customization and Extensibility

Reviewer: (nit / optional fix)
```suggestion
### Customization and extensibility
```

Kubernetes architecture allows for significant customization:

> - Custom schedulers can be deployed alongside or instead of the default scheduler.

Reviewer 2: Improved clarity
```suggestion
- Custom schedulers can be deployed to work alongside the default Kubernetes scheduler or to replace it entirely.
```

Author: Thank you for the rewrite @network-charles, I think that is much clearer!

  The architectural concepts behind Kubernetes.
---

> {{< figure src="/images/docs/kubernetes-cluster-architecture.svg" alt="Components of Kubernetes" caption="**Note:** This diagram presents an example reference architecture of a Kubernetes cluster. The actual distribution of components can vary based on specific cluster setups and requirements." class="diagram-large" >}}

Reviewer: Try setting a title as well:
```suggestion
{{< figure src="/images/docs/kubernetes-cluster-architecture.svg" alt="The control plane (kube-apiserver, etcd, kube-controller-manager, kube-scheduler) and several nodes. Each node is running a kubelet and kube-proxy." 
title="Kubernetes cluster components"
caption="**Note:** This diagram presents an example reference architecture for a Kubernetes cluster. The actual distribution of components can vary based on specific cluster setups and requirements." class="diagram-large" >}}
```

That way, the caption is more obviously a secondary comment about the diagram.

> ### Understanding Cluster Component Distribution

Reviewer: ```suggestion
## Components of a Kubernetes cluster
```

> While the diagram provides a simplified overview, it's important to understand the flexibility of Kubernetes architecture:

Reviewer: ```suggestion
A Kubernetes cluster consists of a control plane and some nodes. The diagram shows an example of what's involved.
```

- **Control Plane Components:** The control plane refers to a collection of critical components that manage the cluster's core functionality. These include the API server, scheduler, controller manager, and etcd (for cluster state storage).

- **Node Components:** Both control plane nodes and worker nodes typically run kubelet (the node agent) and kube-proxy (for network proxying). These components are essential for node operation but are not always considered part of the core control plane.

- **Deployment Variations:** In production environments, there are many possible setups:
  - Control plane components might run on dedicated nodes or be distributed across cluster nodes.
  - They could be deployed as static Pods, managed by systemd, or handled by cluster management tools.
  - In some cases, especially with managed Kubernetes services, the control plane might be abstracted away from direct user management.

> - **Workload Placement:** Depending on the cluster's size and configuration, you might deploy workloads on control plane nodes, especially for cluster management tasks, monitoring, or trusted applications.

Reviewer: Consider using a [description list](https://www.google.com/search?q=markdown+description+list) here. Using a description list helps a bit with accessbility and also makes it easier for maintainers to restyle the site if we ever want to.

Author: Thanks for the suggestion, I've tried to use description lists where applicable
